### ISSUE TITLE
torch.where backward propagates NaN gradient from inactive sqrt branch

### ISSUE THREAD

**Issue by ALinrunrun** [pytorch/pytorch#187297]
### 🐛 Describe the bug

`torch.autograd` produces `NaN` gradients for a `torch.where` expression when the inactive branch contains `torch.sqrt` evaluated on negative inputs.

In the example below, all input values are negative, so the selected branch is the identity branch `xt`. Therefore, the output is simply `xt.sum()`, and the gradient with respect to each input element should be `1.0`. However, PyTorch returns `nan` gradients, apparently due to the inactive `torch.sqrt(xt)` branch.

This issue reproduces on CPU, so CUDA availability does not affect the result.

```
import torch

xt = torch.tensor([-1.0, -4.0], dtype=torch.float32, requires_grad=True)

y = torch.where(xt > 0, torch.sqrt(xt), xt).sum()
y.backward()

actual = xt.grad
expected = torch.tensor([1.0, 1.0], dtype=torch.float32)

print("actual:")
print(actual)
print("expected:")
print(expected)
```

### Actual vs expected result

```text
actual:
tensor([nan, nan])

expected:
tensor([1., 1.])
```

The discrepancy occurs because the selected branch is `xt` for all elements, so the gradient should be finite and equal to `1.0`. Instead, `torch.autograd` returns `nan` gradients, seemingly propagated from the inactive `sqrt` branch.

### Versions

PyTorch version: 2.12.0+cu130
Is debug build: False
CUDA used to build PyTorch: 13.0
ROCM used to build PyTorch: N/A

OS: Ubuntu 24.04.3 LTS (x86_64)
Python version: 3.13.13 | packaged by Anaconda, Inc.
Python platform: Linux x86_64

Is CUDA available: False
GPU models:
  GPU 0: NVIDIA RTX 6000 Ada Generation
  GPU 1: NVIDIA RTX 6000 Ada Generation
Nvidia driver version: 570.211.01

CPU:
  Model name: AMD Ryzen Threadripper PRO 7985WX 64-Cores
  CPU(s): 128

Versions of relevant libraries:
[pip3] numpy==2.4.6
[pip3] torch==2.12.0
[pip3] triton==3.7.0

cc @ezyang @albanD @gqchen @nikitaved @soulitzer @Varal7 @bobrenjc93

**Comment by Xrenya** [comment]
I think you could resolve it:
```python
y = torch.where(xt > 0, torch.sqrt(xt), xt).sum()
```
to this:
```python
y = torch.where(xt > 0, torch.sqrt(torch.clamp(xt, min=0)), xt).sum()

# or another approach
positive_xt = torch.where(xt > 0, xt, torch.ones_like(xt))
y = torch.where(xt > 0, torch.sqrt(positive_xt), xt).sum()
```
`torch.sqrt(xt)` - give `NaN` since all negative, other branch `1`

**Comment by soulitzer** [comment]
https://github.com/pytorch/pytorch/issues/4132
